leases/tasks/send_invoices.py

```diff
@@ -0,0 +1,231 @@
+        lease_year = current_date.year - 1
+    else:  # (2)
+        lease_year = current_date.year
+
+    # Filter leases from the upcoming season
+    future_leases = BerthLease.objects.filter(
+        start_date__year__gt=lease_year,
+        berth=OuterRef("berth"),
+        customer=OuterRef("customer"),
+    )
+
+    # Exclude leases that have already been assigned to the same customer and berth on the future
+    leases: QuerySet = BerthLease.objects.exclude(
+        Exists(future_leases.values("pk"))
+    ).filter(
+        # Only allow leases that are auto-renewing and have been paid
+        renew_automatically=True,
+        status=LeaseStatus.PAID,
+        start_date__year=lease_year,
+        end_date__year=lease_year,
+    )
+
+    return leases
+
+
+def create_new_lease(
+    lease: Union[BerthLease, WinterStorageLease], start_date: date, end_date: date
+) -> Union[BerthLease, WinterStorageLease]:
+    """Create a new lease instance"""
+
+    # Blank the PK to signal that a new instance has to be created
+    lease.pk = None
+
+    # Manually need to set creating because of the checks performed on clean
+    lease._state.adding = True
+
+    # Update the dates for the lease for the next season
+    lease.start_date = start_date
+    lease.end_date = end_date
+    lease.save()
+
+    return lease
+
+
+def send_berth_invoices(request: HttpRequest, due_date: date = None) -> Dict[str, List]:
+    return send_invoices(
+        request,
+        BERTH_LEASES,
+        season_start=calculate_season_start_date(),
+        season_end=calculate_season_end_date(),
+        due_date=due_date,
+    )
+
+
+def send_winter_storage_invoices(
+    request: HttpRequest, due_date: date = None
+) -> Dict[str, List]:
+    raise NotImplementedError
+
+
+@transaction.atomic
+def send_invoices(
+    request: HttpRequest,
+    lease_type: str,
+    season_start: date,
+    season_end: date,
+    due_date: date = None,
+) -> Dict[str, List]:  # noqa: C901
+    successful_orders: List[UUID] = []
+    failed_orders: List[Dict[UUID, str]] = []
+    failed_leases: List[Dict[UUID, str]] = []
+
+    # Default the due date to 14 days from the date when the task is executed
+    if not due_date:
+        due_date = today() + relativedelta(days=14)
+
+    leases = get_valid_leases(season_start, lease_type)
+
+    orders: List[Order] = []
+
+    for lease in leases:
+        lease = create_new_lease(lease, season_start, season_end)
+
+        # Get the associated product
+        product = get_product(lease)
+
+        # If no product is found, the billing can't be done
+        if not product:
+            fail_lease(lease, _("No suitable product found"), failed_leases)
+            continue
+
+        try:
+            order = Order.objects.create(
+                customer=lease.customer, lease=lease, product=product
+            )
+            orders.append(order)
+        except ValidationError as e:
+            fail_lease(lease, str(e), failed_leases)
```

Comment:
Same as above, what are the admins going to do with the draft lease if this fails? If a draft is left in the database and the send_invoices() is executed again, do I understand correctly that get_valid_berth_leases will exclude this lease from renewal?

Reply:
Maybe log the stacktrace to ease debugging?

Reply:
If `send_invoices` is run again, it ignores _all_ the renewed leases for the upcoming season. Meaning that if the lease is in `ERROR`, it won't try to generate it again.

> do I understand correctly that get_valid_berth_leases will exclude this lease from renewal? Or is this the desired result?

Yes to both. On the `Kausilaskut` section on the Admin UI they will be able to check which leases are in error state and handle them. As you pointed out on the previous one, we'll store the error message on the `comment` field so they/we know what was the issue.